Prepare the proposal for release

ThunderDeliverer opened the pull request:
Everything not essential to the proposal being readable and/or runnable has been removed to prepare the proposal for release.

ThunderDeliverer committed:
Prepare the proposal for release

Everything not essential to the proposal being readable and/or runnable
has been removed to prepare the proposal for release.

steven2308 reviewed (approved):
I thought we were removing this on the fork instead, but I was able to run the tests so LFG!

ThunderDeliverer committed:
Update to the state of the current proposal

steven2308 reviewed (commented):
I know it's draft, but remember this now needs an extra level of collection on most methods.

ThunderDeliverer commented:
I just meant to do all of the changes on this branch, but might make more sense to merge with master, yes 🙈

ThunderDeliverer committed:
Rewrite the proposal

Rewrote the proposal to a public good repository from an extension.

ThunderDeliverer committed:
Minor fixes

steven2308 commented on assets/contracts/mocks/ERC721Mock.sol:
```diff
@@ -0,0 +1,20 @@
+ * @title ERC721Mock
+ * Used for tests
+ */
+contract ERC721Mock is ERC721, Emotable {
```
This is surely broken, it's inheriting a non imported contract.
Which doesn't need to inherit anymore anyway

ThunderDeliverer replied:
It seems that GitHub has once again not showed all of the changes 🤔  This was already removed in the following commit.

steven2308 commented on assets/test/emotableRepository.ts:
```diff
@@ -0,0 +1,101 @@
+
+  it('can support IERC721', async function () {
+    expect(await repository.supportsInterface('0x80ac58cd')).to.equal(true);
+  });
```
Why should it support 721 anymore?

ThunderDeliverer replied:
Same thing here. This was changed to check for 165 support.

ThunderDeliverer committed:
Remove 721 from the required EIPs

As this proposal now works independently of ERC-721 smart contracts, we
should remove it from the required EIPs.

ThunderDeliverer commented:
Additional thing to note is that ERC-721 is no longer required for our proposal to work. I'm thinking that we cold even mention that this works for ERC-1155 as well.. 🤔

steven2308 commented:
> Additional thing to note is that ERC-721 is no longer required for our proposal to work. I'm thinking that we cold even mention that this works for ERC-1155 as well.. thinking

Not really sure, I think it would need an additional ID to identify the collection if using 1155.

steven2308 commented on README.md:
```diff
@@ -39,58 +39,94 @@ Current solutions are either proprietary or off-chain and therefore subject to m
+    /**
+     * @notice Used to get the information on whether the specified address has used a specific emoji on a specific
+     *  token.
+     * @dev As storing a uint256 is cheaper than a bool, we use 1 for true and 0 for false.
```
```suggestion
```

steven2308 commented on README.md:
```diff
@@ -99,30 +135,35 @@ Designing the proposal, we considered the following questions:
 The proposal only accepts the Unicode identifier which is a `bytes4` value. This means that while we encourage implementers to add the reactions using standardized emojis, the values not covered by the Unicode standard can be used for custom emotes. The only drawback being that the interface displaying the reactions will have to know what kind of image to render and such additions will probably b ...
 2. **Should the proposal use emojis to relay the impressions of NFTs or some other method?**\
 The impressions could have been done using user-supplied strings or numeric values, yet we decided to use emojis since they are a well established mean of relaying impressions and emotions.
+3. **Should the proposal establish an emotable estension or a common-good repository?**\
```
```suggestion
3. **Should the proposal establish an emotable extension or a common-good repository?**\
```

steven2308 commented on README.md:
```diff
@@ -99,30 +135,35 @@ Designing the proposal, we considered the following questions:
+3. **Should the proposal establish an emotable estension or a common-good repository?**\
+Initially we set out to create an emotable extension to be used with any ERC-721 compilant tokens. However, we realized that the proposal would be more useful if it was a common-good repository of emotable tokens. This way, the tokens that can be reacted to are not only the new ones but also the old ones that have been around since before the proposal.
+4. **Should we use a pre-determined address for the repository or just standardize the interface and force potential users to search for a deployed repository?**\
+We decided to use a pre-determined address for the repository since it would be easier for users to find and use it. This way, the users would not have to search for a deployed repository and they would not have to worry about the repository being deployed on a different address than the one they are used to.\
```
This feels a bit vague or repetitive.

ThunderDeliverer replied:
I just merged the two.

steven2308 commented on assets/contracts/EmotableRepository.sol:
```diff
@@ -0,0 +1,86 @@
+        bool state
+    ) external override {
+        _emote(collection, tokenId, emoji, state);
+    }
```
I guess it's too late now since bytes would change. But since this is a standalone it makes no sense to have an external and an internal function. This will make every operation a bit more expensive I think.

ThunderDeliverer replied:
It's not really an issue.. I removed the internal function and recalculated the address 👌

steven2308 commented on assets/contracts/EmotableRepository.sol:
```diff
@@ -0,0 +1,86 @@
+        uint256 tokenId,
+        bytes4 emoji,
+        bool state
+    ) internal virtual {}
```
These hooks don't make sense at all, they were only useful as an extension.
I don't think we can pass the proposal with those.

ThunderDeliverer replied:
I agree, I removed this.

steven2308 reviewed (changes requested):
Really nice job with the description and code changes. 
Sorry I did not catch the hooks and double emote function before, but I do think we need to change that :disappointed:

ThunderDeliverer committed:
Apply changes based on code review

Applied the changes based on code review.

ThunderDeliverer committed:
Mention the compatibility with ERC-1155 in the Abstract

steven2308 commented on assets/contracts/EmotableRepository.sol:
```diff
@@ -0,0 +1,63 @@
+            emoji
+        ] == 1;
+        if (currentVal != state) {
+            _beforeEmote(collection, tokenId, emoji, state);
```
```suggestion
```

steven2308 commented on assets/contracts/EmotableRepository.sol:
```diff
@@ -0,0 +1,63 @@
+                ? 1
+                : 0;
+            emit Emoted(msg.sender, collection, tokenId, emoji, state);
+            _afterEmote(collection, tokenId, emoji, state);
```
```suggestion
```

steven2308 reviewed (approved):
Let's remove the hook calls and make sure tests are fine. 
Other than that looks good to go!

ThunderDeliverer committed:
Apply suggestions from code review

Co-authored-by: Steven Pineda <steven2308@outlook.com>

ThunderDeliverer committed:
Change visibility of emote in EmoteRepository

steven2308 reviewed (commented):
Does it really need the override?

ThunderDeliverer committed:
Remove the unnecesary override

ThunderDeliverer commented:
> Does it really need the override?

No, I've been dealing with a weird error when running the tests and this was one of the solutions I tried. It sneaked past as it didn't impact the operation.

ThunderDeliverer committed:
Tag address as TBA

ThunderDeliverer committed:
Add multi operations and presigned operations

Multi-setters and getters were added in order to provide more
versatility to the contract.
Presigned operations were added in order to allow users to sign
operations and send them to a third party to execute them.

ThunderDeliverer committed:
Add tests and minor modifications to repository

Added tests for the EmotableRepository contract and modified the
contract to correctly recover signer address from the ECDSA signature.

ThunderDeliverer committed:
Run prettier

ThunderDeliverer committed:
Update the interface in the specification to reflect the new methods

ThunderDeliverer committed:
Update the rest of the proposal to be in line with currently published repository

ThunderDeliverer committed:
Update proposal Rationale and Sacurity considerations

ThunderDeliverer committed:
Extend Rationale with a note on chain ID use

ThunderDeliverer committed:
Apply suggestions from code review

Co-authored-by: Steven Pineda <steven2308@outlook.com>

ThunderDeliverer committed:
Apply changes based on PR discussion

ThunderDeliverer committed:
Update the proposal with the latest changes

ThunderDeliverer committed:
Merge pull request #8 from rmrk-team/add-multi-operations

Add multi operations
